### ISSUE TITLE
fatal error: concurrent map writes when multiple backends

### ISSUE THREAD

**Issue by bartoszgolek** [devopsfaith/krakend#105]
```krakend[4734]: fatal error: concurrent map writes
krakend[4734]: goroutine 1053 [running]:
krakend[4734]: runtime.throw(0xc8a990, 0x15)
krakend[4734]: /var/opt/go/1.10/src/runtime/panic.go:619 +0x81 fp=0xc42061d920 sp=0xc42061d900 pc=0x42c921
krakend[4734]: runtime.mapassign_faststr(0xb74440, 0xc4205adda0, 0xc84b14, 0xd, 0xc4201eacb0)
krakend[4734]: /var/opt/go/1.10/src/runtime/hashmap_fast.go:703 +0x3e9 fp=0xc42061d990 sp=0xc42061d920 pc=0x40d529
krakend[4734]: ioki.pl/krakend-ce/oauth1.newMiddleware.func1.1(0xd225c0, 0xc4201a0840, 0xc4205ca140, 0xc4205f85a6, 0xc4205f85c0, 0xc42064ac78)
krakend[4734]: /home/bartoszgolek/go/src/ioki.pl/krakend-ce/oauth1/backend.go:35 +0x12c fp=0xc42061dbc0 sp=0xc42061d990 pc=0x96f96c
krakend[4734]: ioki.pl/krakend-ce/vendor/github.com/devopsfaith/krakend-metrics.NewProxyMiddleware.func1.1(0xd225c0, 0xc4201a0840, 0xc4205ca140, 0x0, 0x0, 0x50)
krakend[4734]: /home/bartoszgolek/go/src/ioki.pl/krakend-ce/vendor/github.com/devopsfaith/krakend-metrics/proxy.go:56 +0xd9 fp=0xc42061dc88 sp=0xc42061dbc0 pc=0x95de49
krakend[4734]: ioki.pl/krakend-ce/vendor/github.com/devopsfaith/krakend/proxy.newLoadBalancedMiddleware.func1.1(0xd225c0, 0xc4201a0840, 0xc4205ca0f0, 0x412468, 0x10, 0xba0580)
krakend[4734]: /home/bartoszgolek/go/src/ioki.pl/krakend-ce/vendor/github.com/devopsfaith/krakend/proxy/balancing.go:59 +0x304 fp=0xc42061ddc0 sp=0xc42061dc88 pc=0x6dba54
krakend[4734]: ioki.pl/krakend-ce/vendor/github.com/devopsfaith/krakend/proxy.NewRequestBuilderMiddleware.func1.1(0xd225c0, 0xc4201a0840, 0xc42049ca50, 0xc4201a0840, 0xc4203620d0, 0xc4201282d0)
krakend[4734]: /home/bartoszgolek/go/src/ioki.pl/krakend-ce/vendor/github.com/devopsfaith/krakend/proxy/http.go:87 +0x1df fp=0xc42061deb8 sp=0xc42061ddc0 pc=0x6dcd3f
krakend[4734]: ioki.pl/krakend-ce/vendor/github.com/devopsfaith/krakend/proxy.requestPart(0xd22640, 0xc420020f60, 0xc4202dd770, 0xc42049ca50, 0xc420020fc0, 0xc420021020)
krakend[4734]: /home/bartoszgolek/go/src/ioki.pl/krakend-ce/vendor/github.com/devopsfaith/krakend/proxy/merging.go:61 +0x84 fp=0xc42061dfb0 sp=0xc42061deb8 pc=0x6dad14
krakend[4734]: runtime.goexit()
krakend[4734]: /var/opt/go/1.10/src/runtime/asm_amd64.s:2361 +0x1 fp=0xc42061dfb8 sp=0xc42061dfb0 pc=0x459921
krakend[4734]: created by ioki.pl/krakend-ce/vendor/github.com/devopsfaith/krakend/proxy.NewMergeDataMiddleware.func1.1
```

I think it is connected to: 
`https://github.com/devopsfaith/krakend/pull/98/commits/4fbd601e5f5f7799cd0e693f7649d7f0ee0db076`

I've found that better solutino would be:
`https://github.com/devopsfaith/krakend/commit/079376adbc655473d3eabd12e6c91581bdb77597`

**Comment by bartoszgolek** [comment]
OAuth1 Code responsible for lines that throws error is: 
```header, err := createOauth1Header(cfg, *request)
			request.Headers[authorizationHeaderKey] = []string{
				header,
			}

			resp, err := next[0](ctx, request)```

**Comment by alombarte** [comment]
How could we reproduce this?

**Comment by bartoszgolek** [comment]
Could be tough, it is some race I think. It not occurs even.

**Comment by alombarte** [comment]
I need something actionable to start exploring this issue :(

**Comment by bartoszgolek** [comment]
I know, it is the tough part of multithreading. I saw this error few times in logs, but then after start service again it works.

**Comment by bartoszgolek** [comment]
I think it will be possible to reproduce this using martian too. I will try tomorrow, but because IMO it is a thread race it will not be reproducible at each request

**Comment by kpacha** [comment]
hi, @bartoszgolek 

after inspecting the stacktrace, I think it crashes before executing the `proxy.NewHTTPProxyDetailed` function, so you're still dealing with a `proxy.Response`.

that's why I think this is the problem: https://github.com/devopsfaith/krakend/blob/master/proxy/request.go#L46

It's the same bug you found in the backend layer, but this time in the proxy.

I'm pushing a fix right now. Let me know if it doesn't work

**Comment by bartoszgolek** [comment]
No problem with that, I don't need to modify parameters. Just suggested it for consistency.

**Comment by kpacha** [comment]
@bartoszgolek , I've being doing some benchmarks and tests and I think the way we were exploring is not the right one.

If the argument is "all proxy middlewares should be able to modify either the `proxy.Request` and the `proxy.Response` received by pointer without thinking about concurrency", I'm afraid we'll end hurting the global performance of the framework because we should 'hard clone' both entities at almost every `proxy.Proxy` function.

On the other hand, the same statement is safe and sound if we restrict it to the `backend` layer because the entities has already been hard cloned.

So, my position is:

**we should keep the `proxy.Request.Clone()` as weak as it is right now and document it properly.**

Also, we should let know the middleware authors that **they must deal with the hard copy of the entity (either the `proxy.Request` and/or the `proxy.Response`) received by pointer if they want to change them at a given `proxy.Proxy` function outside the the backend layer**

Bottomline: you should 'hard clone' the `proxy.Request` in your custom middleware

**Comment by kpacha** [comment]
One more thing... 

If you tune the name and export the functions you created at https://github.com/devopsfaith/krakend/commit/079376adbc655473d3eabd12e6c91581bdb77597 (maybe something like `CloneRequestHeader(req *Request) map[string][]string` and `CloneRequestParams(req *Request) map[string]string`),  we'll add those functions into the framework so other can reuse them if necessary

Do you think we should also expose a full `CloneRequest(req *Request) *Request` function overwriting the `*Request` returned by the `req.Clone()` method with the cloned headers and params?

**Comment by bartoszgolek** [comment]
@kpacha I realy appreciate Your answer, but I would like to argue with it.
I have few point why:
1. If every middleware tah needs to modify headers will copy them (to ensure avoiding this panic) the performance will be broken
2. The need of deep clone is required only if multiple backends exists for 1 endpoint and when headers are modified (maybe some option in cofig)
3. After [fix](https://github.com/devopsfaith/krakend/pull/98/commits/4fbd601e5f5f7799cd0e693f7649d7f0ee0db076) we are doing it anyway (and for single backend endpoints too) but IMO too late.

So my thought is that we need to move cloning to `github.com/devopsfaith/krakend/proxy/http.go:84` not to clone method because this affects loadbalancing too. Then we can revert previous fix (without change in performance (CPU usage) and with cost of few miliseconds (per request) of additional memory usage (when executing all backends, in oposit to current when executing request).

**Comment by bartoszgolek** [comment]
Furthermore, I cant reproduce it, but looking at MartianCode, there is a possibility of such panic when using  marttian too. It uses and modifies headers given as pointer.

**Comment by kpacha** [comment]
@bartoszgolek , after some time 'away', here you have my followup comment regarding your points:

1. I think my wording didn't help my point, let me rephrase: If we don't limit the 'hard clone' to the middlewares actually making changes to the parts of the `proxy.Request` that could affect other concurrent copies, we'd impact on all the built pipes, not just the ones using the middlewares requesting the 'extra' modifications. The headers or the arguments could be changed in the latter layer (backend) with no effort or cost.
2. 100% agree. That's why I'd rather require these middlewares to actually do the deep cloning than enforcing that strategy at the 'merging' and 'concurrent' middlewares at every pipe. Another alternative is to set-up your middleware in the backend layer, when no concurrent processing is done and everything has been already hard copied.
3. The fix you're are linking actually copies the headers when the `proxy.Request` is converted to a `http.Request` instead of copying the header map (and using it on the generated `http.Request`). The main difference here is that this middleware is executed concurrently once per backend registered in a pipe. So if one endpoint hits 2 backends, the original `proxy.Request` will be used in both goroutines to call a deep clone of the headers when their time to be transformed into a `http.Request` arrives. If we place a middleware modifying the `proxy.Request` after the merging middleware but before the backend layer, its changes on one goroutine will affect others unless it hard clones the entity it is ready to modify before passing it to the next element in the chain (in order to avoid the panics and possible side-effects).

> So my thought is that we need to move cloning to github.com/devopsfaith/krakend/proxy/http.go:84 not to clone method because this affects loadbalancing too. Then we can revert previous fix (without change in performance (CPU usage) and with cost of few miliseconds (per request) of additional memory usage (when executing all backends, in oposit to current when executing request).
> 
> Maybe it is a good idea to add method to proxy.Request named DeepClone

That was almost my point: leave the clone as it is and expose two package functions as helpers for 'hard cloning' the headers and/or the params of a `proxy.Request` for "safe" modifications in the proxy layer, before the backend. In the following picture, the 'backend layer' is where the three lines at the right end ![sample pipe](http://www.krakend.io/images/KrakendFlow.png) 

> Furthermore, I cant reproduce it, but looking at MartianCode, there is a possibility of such panic when using martian too. It uses and modifies headers given as pointer.

As said, every execution at the backend layer is 'thread-safe' because the goroutines of all the backends related to the same pipe execution do not share a thing. At this stage of the pipe, every middleware in the pipe is executed synchronously, from the outer layer to the inner one and all the way back.

Now the question I have is: should we ever clone the headers or the params in any middleware included in the `proxy` package of the framework? Maybe we should go with the 'performance first' moto here too and require every middleware (even at the backend layer) to the deal with the hard clone if required, so the performance of the 'leaner' pipes will be as smooth as possible.

**Comment by bartoszgolek** [comment]
> As said, every execution at the backend layer is 'thread-safe' because the goroutines of all the backends related to the same pipe execution do not share a thing.
If I understand correctly, I cant agree. Headers and params are shared and this is why I'm getting `concurrent map writes` panic when adding header in my custom registered backend middleware.

I'm aware of additional performance cost of cloning headers, but still I think that every code run in goroutine `github.com/devopsfaith/krakend/proxy/merging.go:33` should be a deep clone to avoid concurency issues. So, refering to p2, is it ok to add clone of Headers (and maybe Params) before calling next in `github.com/devopsfaith/krakend/proxy/merging.go:61`?

**Comment by kpacha** [comment]
hi, @bartoszgolek . Please, check the PR #125 if you want to keep your middleware in the 'not thread-safe' part of the pipe
